docs/readmes/proposals/p010_vendor_neutral_dp.md

```diff
@@ -416,7 +456,250 @@ scope of work.
+
+  On the way back from SAS the requests will be put back into the PC
+  and distributed across CBSDs that sent them.
+
+  There will be one PC for each protocol that the DP needs to handle.
+
+  Initially this means only Domain Proxy protocol, but the idea behind
+  Protocol Controllers is for them to be written in such a manner that
+  having one of them in place, it should be straightforward to add
+  another one (it can be achieved by having a ProtocolController abstract class
+  to provide structure for particular PC implementations).
+  
+  To scale the system better we are decoupling PC (responsible for protocol)
+  from Radio controller (responsible for the business domain).
+  Additional benefit is maintaining all of the protocol customizations
+  and tricks at the PC level and making the PC layer scalable
+  according to the number of eNBs of each type.
+
+* **Radio Controller (RC)**
+  
+  RC is responsible for the CBSD to SAS communication
+  abstracted from vendor specific protocols.
+  
+  RC on one hand is responsible for placing the requests coming from PCs in the queue
+  and on the other for segregating the responses coming from Configuration Controller
+  and sending them back to the appropriate PC.
+
+* **Configuration Controller (CC)**
+
+  Configuration Controller is responsible for polling requests periodically from the queue,
+  aggregating them into one request in format understandable by SAS,
+  posting them over to SAS
+  and once a response is received, passing it on to RC.
+
+* **Message Queue**
+
+  RabbitMQ instance for storage of requests received by the RC
+  allowing them to be polled by a consumer.
+  Requests for which acknowledgement is received upon polling
+  will be deleted from the queue,
+  while the ones without acknowledgement will be requeued.
+
+* **Database (optional, to be implemented at stage 3)**
+
+  Relational DB storing the state of the DP.
+
+  DB allows to keep track of grants and CBSDs configuration limiting redundant CBSD - SAS connections.
+
+## Component Interactions
+
+* **Protocol Controller**
+
+  PC will pass the requests it receives from CBSDs to the RC over HTTPS.
+  
+  PC will pass the responses that are coming from RC back to the respective CBSDs
+  over the protocol that they are able to understand and the PC is implementing.
+
+* **Radio Controller**
+
+  RC receives requests from the PC over HTTPs and puts them in
+  the queue over AMQP.
+
+  When the response for the request comes back from CC,
+  RC will forward it over HTTPS to the appropriate PC.
+
+* **Configuration Controller**
+
+  CC communicates with the queue over AMQP for request polling.
+  
+  CC communicates with SAS (external) over HTTPS for posting bulked requests and with the RC over HTTPS for forwarding the SAS response.
+
+* **Message Queue**
+
+  Treating RC as the publisher and CC as the consumer.
+
+* **Database**
+
+  Exchanging information about CBSDs and grants with the CC.
+
+## Solution scaling
```

Comment:
Can we add a section on how we can graceful degradation. I wonder if we have a situation where configuration controller unable to reach SAS controller, in that case the queue will end building up and when the SAS controller comes back up it might not be able to handle the load and will again start say rate limiting the requests and most of our requests will end up in errors. In such cases instead of keeping the CBSDs in dark, might be useful if we can provide feedback that system isn't available and for CBSDs to pause their requests.

Reply:
@karthiksubraveti I think we had a similar internal discussion today if I'm following. We can decouple the state machines between CBSD and DP and DP and SAS for any given CBSD. We can have asynchronous messaging going on between CBSD and DP and DP and SAS. The only thing that has to be linked to true SAS state is the grant messaging that reaches the "Authorized" state. Prior to that the radio will be off and we can signal as needed to manage the CBSD state.